src/type/mod.rs

```diff
@@ -0,0 +1,232 @@
+// Copyright 2019
+//     by  Centrality Investments Ltd.
+//     and Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::tm_std::*;
+
+use crate::{
+	form::{CompactForm, Form, MetaForm},
+	IntoCompact, MetaType, Metadata, Registry,
+};
+use derive_more::From;
+use serde::Serialize;
+
+mod composite;
+mod fields;
+mod product;
+mod sum;
+
+pub use self::{composite::*, fields::*, product::*, sum::*};
+
+/// Implementors return their meta type identifiers.
+pub trait HasType {
+	/// Returns the static type identifier for `Self`.
+	// todo: [AJ] good name for this? r#type() perhaps?
+	fn get_type() -> Type;
+}
+
+/// A type identifier.
+///
+/// This uniquely identifies types and can be used to refer to type definitions.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, From, Debug, Serialize)]
+#[serde(bound = "
+	F::Type: Serialize,
+	F::IndirectType: Serialize
+")]
+#[serde(rename_all = "camelCase")]
+pub enum Type<F: Form = MetaForm> {
+	/// A product type (e.g. a struct)
+	Product(TypeComposite<TypeProduct<F>, F>),
+	/// A sum type (e.g. an enum)
+	Sum(TypeComposite<TypeSum<F>, F>),
+	/// A slice type with runtime known length.
+	Slice(TypeSlice<F>),
+	/// An array type with compile-time known length.
+	Array(TypeArray<F>),
+	/// A tuple type.
+	Tuple(TypeTuple<F>),
+	/// A Rust primitive type.
+	Primitive(TypePrimitive),
+}
+
+impl IntoCompact for Type {
+	type Output = Type<CompactForm>;
+
+	fn into_compact(self, registry: &mut Registry) -> Self::Output {
+		match self {
+			Type::Product(product) => product.into_compact(registry).into(),
+			Type::Sum(sum) => sum.into_compact(registry).into(),
+			Type::Slice(slice) => slice.into_compact(registry).into(),
+			Type::Array(array) => array.into_compact(registry).into(),
+			Type::Tuple(tuple) => tuple.into_compact(registry).into(),
+			Type::Primitive(primitive) => primitive.into(),
+		}
+	}
+}
+
+pub fn product_type<T, P>(name: &'static str, namespace: Namespace, type_params: P, def: T) -> Type
+where
+	T: Into<TypeProduct>,
+	P: IntoIterator<Item = MetaType>,
+{
+	Type::Product(TypeComposite::new(name, namespace, type_params, def.into()))
+}
+
+pub fn sum_type<T, P>(name: &'static str, namespace: Namespace, type_params: P, def: T) -> Type
+where
+	T: Into<TypeSum>,
+	P: IntoIterator<Item = MetaType>,
+{
+	Type::Sum(TypeComposite::new(name, namespace, type_params, def.into()))
+}
```

Comment:
These two could probably be inherent methods of `Type` and return respective builders that respect the different properties of tuple-structs (unnamed fields), structs (named fields), c-like enums (discriminants, no data), variants (no discriminants, maybe data) etc.